### PR TITLE
Fix 4520 princess only fires aero tag

### LINKED ISSUE #4520: Princess never uses other weapons if TAG pod is loaded or present on aircraft
If Princess is given control of a fighter, conventional or aerospace, that is equipped with TAG it will only use TAG and never make any other attacks.  Aircraft using TAG cannot make weapon attacks, so there should be some kind of decision making between the TAG and it's other weapons.

Under ideal circumstances, Princess would take into account incoming precision weapons (e.g. homing Arrow IV artillery) and LRM-armed ground units within a certain range of the TAG'd target when deciding whether to use TAG on a target or make a standard weapon attack (e.g. no friendly LRMs within 14 hexes for indirect fire, don't bother with the TAG).

Attached 49.12 save game for demonstration and testing:
[Princess_air_TAG.sav.gz](https://github.com/MegaMek/megamek/files/11717907/Princess_air_TAG.sav.gz)

The Thunderbird for Unknown Hostiles has a TAG pod, weapons (obviously) and several air-launched Arrow IV.  It will never use anything but the TAG despite mothing to take advantage of .

### PR DESCRIPTION
**Pull request by Sleet01**
This change updates a few things:

1. don't allow Aerospace units to fire TAG if there's no utility in it.
2. let Princess calculate incoming guided fire for the turn, and use this to calculate Aero TAG utility.
3. move part of the "has usable weapons" check out into `compute` as a generic `computeTotalDamage` function, and include bombs in the computation.

Since the implementation of step 2 is pretty simplistic, that part can be excised for a later patch without affecting the fix to this issue.

In short, for each Aerospace with TAG that Princess controls, Princess will now calculate and cache a list of 
A) all incoming guided artillery impacting this turn;
B) all Laser-Guided Bombs on other units on the board;
C) all Indirect Fire-capable weapons on all other (non-Aerospace) units on the board;

these lists will be used to compute the maximum possible damage deriving from using a given Aerospace unit's TAG rather than firing its weapons normally.  The Aero unit will attempt to TAG if its expected damage is equal to or less than the potential damage from incoming guided munitions this turn.  Otherwise, TAG will not be used.

NOTE: This calculation does _not_ take into account the distance from a given TAG target to each incoming munition's target point (needed for precise assessment of e.g. Homing Arrow IV or Copperhead rounds).  It also does not take into account whether prospective targets are even within firing range of Indirect-Fire-capable friendly ground units.  However, adding these checks to `computeGuidedWeapons` would not cause much additional computation time overall because each list is calculated once and then cached for later re-use by the turn-planning code.

tl;dr: fix works, and Princess will be _slightly_ smarter about using TAG.

Testing:
- Tested with Princess-controlled Bomber ASFs mounting normal weapons, guided bomb munitions, and TAG (both as weapons and as "bomb" pods).
-- Also included off-board Copperhead and Homing Arrow IV units.
- Ran all MegaMek unit tests.

close #4520

### CODE REVIEW

**Comment by Sleet01** [comment]
I'll have another commit tomorrow with fixes for the above, and hopefully filtering out homing attacks by distance.

**Comment by Sleet01** [comment]
A fun turn testing with this code:

image
image

Teammate, a friendly Princess, detected two previously-fired Homing AIVs and a bunch of LG bombs nearby, so used the relatively poorly-armed Centurions to TAG for the Homing rounds and the Sabutai's dive bomb attack, as well spotting for indirect Thunderbolt fire (not shown).

**Comment by HammerGS** [comment]
Awesome to see.

File: megamek/src/megamek/common/Entity.java [hunk 1]
```diff
@@ -868,30 +875,19 @@ protected void initMilitary() {
     }
 
     protected boolean hasViableWeapons() {
-        int totalDmg = 0;
+        int totalDmg = Compute.computeTotalDamage(getTotalWeaponList());
```
[the remark this answers is not in the document]
Comment (NickAragua) [reply megamek/src/megamek/common/Entity.java:878]: I feel like we could also optimize this for when we reach more than 5 total damage, but I guess that's in another loop now.

**Review by NickAragua — COMMENTED** [review]
Some thoughts/feedback.

File: megamek/src/megamek/common/Compute.java [hunk 2]
```diff
@@ -1089,6 +1090,21 @@ public static ToHitData getImmobileMod(Targetable target, int aimingAt, AimingMo
         return null;
     }
 
+    public static boolean isIndirect(WeaponType wtype) {
```
[the remark this answers is not in the document]
Comment (HammerGS) [reply megamek/src/megamek/common/Compute.java:1093]: > (Unrelated aside: how tf do indirect-fire torpedoes work? lol)
What goes up...must come down :)

[the remark this answers is not in the document]
Comment (gsparks3) [reply megamek/src/megamek/common/Compute.java:1093]: >(Unrelated aside: how tf do indirect-fire torpedoes work? lol)

The path still has to be completely through water, but you don't need direct LOS, so you can have them like, snake around a river to hit something that can't see you, AFAIK. Doesn't come up much if both combatants are in open water or the same lake, but could be nasty in a winding river.

[the remark this answers is not in the document]
Comment (mmetke-qumulo) [reply megamek/src/megamek/common/Compute.java:1093]: > Also, don't we already have a way of telling if a weapon is capable of indirect fire as per missile rules?
> 
> (Unrelated aside: how tf do indirect-fire torpedoes work? lol)

I think we can tell if a missile weapon _is in indirect mode_, and I think we have an Indirect flag, but I'm not sure if all weapons that _are_ Indirect (tube arty, arty cannons, Mek Mortars, etc.) support all those checks.

Moving this to WeaponType makes sense.

File: megamek/src/megamek/common/Entity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/Entity.java:878]: I'd need to learn a bit more about Java concurrency; I attempted to put the incoming damage lists in each Entity but ended up with mismatched states and nothing worked so I pushed the list into Princess instead.

It would probably be a good idea to see how much stuff each Entity currently stores, how many times the various calculations get called, and whether that could all be sped up by caching... maybe a separate RFE?

File: megamek/src/megamek/common/Compute.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (NickAragua) [reply megamek/src/megamek/common/Compute.java]: Why static and not just a standard method? Based on the existing code, we're definitely not worried about weapon type being null, and having the method be a part of the instance saves us having to pass an extra object reference around.

File: megamek/src/megamek/common/Entity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (NickAragua) [reply megamek/src/megamek/common/Entity.java:878]: I started caching some things in the CachedEntityState class to optimize performance a little bit. I agree that it's beyond the scope of this PR.

File: megamek/src/megamek/common/Compute.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sleet01) [reply megamek/src/megamek/common/Compute.java]: > Why static and not just a standard method? Based on the existing code, we're definitely not worried about weapon type being null, and having the method be a part of the instance saves us having to pass an extra object reference around.

Just pure laziness.  The most correct thing would be to make it a base class method and only return `true` in the above cases which... might as well.

**Review by NickAragua — APPROVED** [review]
